File: apps/boa/lib/features/notifications/notification_priority.dart
enum NotificationPriority { urgent, today, general }

const Set<String> _urgentTypes = {
  'schedule_incomplete',
  'long_unmanaged_90',
  'unpaid_lapse',
  'reconsult',
  'uncontacted_3days',
};

const Set<String> _todayTypes = {
  'schedule_today',
  'schedule_1hour',
  'birthday',
};

NotificationPriority classifyNotificationPriority(Map<String, dynamic> raw) {
  final type = '${raw['type'] ?? ''}';
  if (_urgentTypes.contains(type)) return NotificationPriority.urgent;
  if (_todayTypes.contains(type)) return NotificationPriority.today;

  final dueAtRaw = raw['dueAt'];
  if (dueAtRaw != null) {
    final dueAt = DateTime.tryParse('$dueAtRaw');
    if (dueAt != null) {
      final now = DateTime.now();
      final due = DateTime(dueAt.year, dueAt.month, dueAt.day);
      final today = DateTime(now.year, now.month, now.day);
      if (!due.isAfter(today)) return NotificationPriority.today;
    }
  }
  return NotificationPriority.general;
}

int notificationPriorityWeight(NotificationPriority priority) {
  switch (priority) {
    case NotificationPriority.urgent:
      return 0;
    case NotificationPriority.today:
      return 1;
    case NotificationPriority.general:
      return 2;
  }
}

bool notificationIsRead(Map<String, dynamic> raw) {
  return raw['isRead'] == true || raw['isRead'] == 1;
}

List<Map<String, dynamic>> sortNotificationsForQueue(Iterable<Map<String, dynamic>> items) {
  final sorted = items.toList(growable: true);
  sorted.sort((a, b) {
    final pa = classifyNotificationPriority(a);
    final pb = classifyNotificationPriority(b);
    final priorityDiff = notificationPriorityWeight(pa) - notificationPriorityWeight(pb);
    if (priorityDiff != 0) return priorityDiff;

    final aRead = notificationIsRead(a);
    final bRead = notificationIsRead(b);
    if (aRead != bRead) return aRead ? 1 : -1;

    final at = DateTime.tryParse('${a['createdAt'] ?? ''}') ?? DateTime.fromMillisecondsSinceEpoch(0);
    final bt = DateTime.tryParse('${b['createdAt'] ?? ''}') ?? DateTime.fromMillisecondsSinceEpoch(0);
    return bt.compareTo(at);
  });
  return sorted;
}

File: client/src/lib/notificationPriority.ts
import {
  NOTIFICATION_TODAY_TYPES,
  NOTIFICATION_URGENT_TYPES,
  type NotificationPriority,
} from "@shared/notificationActionCenter";

export type { NotificationPriority } from "@shared/notificationActionCenter";

const urgentTypes = new Set<string>(NOTIFICATION_URGENT_TYPES);
const todayTypes = new Set<string>(NOTIFICATION_TODAY_TYPES);

export function classifyNotificationPriority(notification: {
  type: string;
  dueAt?: string | Date | null;
}): NotificationPriority {
  if (urgentTypes.has(notification.type)) return "urgent";
  if (todayTypes.has(notification.type)) return "today";

  if (notification.dueAt) {
    const due = new Date(notification.dueAt);
    if (!Number.isNaN(due.getTime())) {
      const now = new Date();
      const dueDateOnly = new Date(
        due.getFullYear(),
        due.getMonth(),
        due.getDate()
      );
      const todayOnly = new Date(
        now.getFullYear(),
        now.getMonth(),
        now.getDate()
      );
      if (dueDateOnly.getTime() <= todayOnly.getTime()) return "today";
    }
  }
  return "general";
}

export function priorityWeight(priority: NotificationPriority): number {
  if (priority === "urgent") return 0;
  if (priority === "today") return 1;
  return 2;
}

export type SortableNotification = {
  type: string;
  dueAt?: string | Date | null;
  createdAt: string | Date;
  isRead?: boolean | number | null;
};

export function sortNotificationsForQueue<T extends SortableNotification>(
  items: T[]
): T[] {
  return [...items].sort((a, b) => {
    const pa = classifyNotificationPriority(a);
    const pb = classifyNotificationPriority(b);
    const weightDiff = priorityWeight(pa) - priorityWeight(pb);
    if (weightDiff !== 0) return weightDiff;

    const unreadDiff = Number(Boolean(a.isRead)) - Number(Boolean(b.isRead));
    if (unreadDiff !== 0) return unreadDiff;

    return new Date(b.createdAt).getTime() - new Date(a.createdAt).getTime();
  });
}
